main.go

```diff
@@ -140,6 +176,100 @@ func main() {
+		key := client.ObjectKey{
+			Name:      hr.Name,
+			Namespace: hr.Namespace,
+		}
+		if err := clientSet.Get(ctx, key, instance); client.IgnoreNotFound(err) != nil {
+			if ctx.Err() == context.DeadlineExceeded {
+				log.Errorf("Timed out polling for the helm release presence")
+			} else {
+				log.Fatalf("Failed to get instance of the helmrelease with %v", err)
+			}
+			return false
+		} else if err != nil {
+			log.Infof("Did not find the helm release: %v, the object is deleted", hr.Name)
+			return true
+		} else {
+			log.Infof("Found the helm release: %v, the object is not deleted", hr.Name)
+			return false
+		}
+	}
+
+	if err := poll(ctx, pollingFunc, time.Second); err != nil {
+		log.Errorf("Failed to poll the deletion of the helm release with %v", err)
+		// The helm release is still not deleted. Force cleanup as the final attempt
+		forceCleanupHelmRelease(clientSet, hr, pollingFunc, timeout)
+	}
+	cancel()
+}
+
+// remove the finalizer from the helm release and wait for it get garbage collected
+// if the helm release is still present, call 'helm uninstall' on it
+func forceCleanupHelmRelease(clientSet client.Client, hr *fluxhelmv2beta1.HelmRelease, poller func() bool, timeout time.Duration) {
+	// Create a new context with timeout
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+	instance := &fluxhelmv2beta1.HelmRelease{}
+	key := client.ObjectKey{
+		Name:      hr.Name,
+		Namespace: hr.Namespace,
+	}
+	if err := clientSet.Get(ctx, key, instance); client.IgnoreNotFound(err) != nil {
+		log.Fatalf("Failed to get instance of the helmrelease with %v", err)
+	} else if err != nil {
+		// The helm release was not found after the first failed cleanup attempt
+		log.Errorf("Did not find the helm release: %v, the object is deleted", hr.Name)
+	} else {
+		log.Infof("Removing the helm release finalizer: %v...", fluxhelmv2beta1.HelmReleaseFinalizer)
+		controllerutil.RemoveFinalizer(instance, fluxhelmv2beta1.HelmReleaseFinalizer)
+		if err := clientSet.Update(ctx, instance); err != nil {
+			log.Fatalf("Failed to update the helmrelease with %v", err)
+		}
+		if err := poll(ctx, poller, time.Second); err != nil {
+			log.Errorf("Failed to poll the deletion of the helm release with %v", err)
+			log.Infof("Uninstalling the release: %v...", hr.Name)
+			settings := helmcli.New()
+			actionConfig := new(helmaction.Configuration)
+			if err := actionConfig.Init(settings.RESTClientGetter(), settings.Namespace(), os.Getenv("HELM_DRIVER"), log.Infof); err != nil {
+				log.Fatalf("Failed to init the helm uninstall action with %v", err)
+			}
+			client := helmaction.NewUninstall(actionConfig)
+			if _, err := client.Run(hr.Name); err != nil {
+				log.Errorf("Failed to uninstall the helm release with %v", err)
+			}
+		}
+	}
+	cancel()
+}
+
 func PollStatus(ctx context.Context, cancel context.CancelFunc, clientSet client.Client, config *rest.Config, identifiers ...object.ObjMetadata) error {
 	defer cancel()
```

Comment:
if we call `defer cancel()` in the `installAction` we can remove this line

Reply:
See my previous reply